File: TyreVibes/Core/Model/TyreAnalysisReportModels.swift
//
//  TyreAnalysisReportModels.swift
//  TyreVibes
//
//  Advanced tyre analysis report data models
//

import Foundation
import SwiftUI
import CoreLocation

// MARK: - Main Report Model
struct TyreAnalysisReport: Codable, Identifiable {
    let id: UUID
    let metadata: ReportMetadata
    let depthAnalysis: DepthAnalysis
    let wearAnalysis: WearAnalysis
    let heatMap: DepthHeatMap
    let remainingLife: RemainingLifeEstimate
    let recommendations: [Recommendation]
    let safetyScore: SafetyScore
    let comparison: ComparisonData?

    var createdAt: Date {
        metadata.timestamp
    }

    init(
        id: UUID = UUID(),
        metadata: ReportMetadata,
        depthAnalysis: DepthAnalysis,
        wearAnalysis: WearAnalysis,
        heatMap: DepthHeatMap,
        remainingLife: RemainingLifeEstimate,
        recommendations: [Recommendation] = [],
        safetyScore: SafetyScore,
        comparison: ComparisonData? = nil
    ) {
        self.id = id
        self.metadata = metadata
        self.depthAnalysis = depthAnalysis
        self.wearAnalysis = wearAnalysis
        self.heatMap = heatMap
        self.remainingLife = remainingLife
        self.recommendations = recommendations
        self.safetyScore = safetyScore
        self.comparison = comparison
    }
}

// MARK: - Report Metadata
struct ReportMetadata: Codable {
    let reportId: String
    let timestamp: Date
    let vehicle: VehicleInfo
    let tyre: TyreInfo
    let location: LocationInfo?
    let weather: WeatherInfo?
    let analysisType: AnalysisType

    enum AnalysisType: String, Codable {
        case quick = "Quick Scan"
        case standard = "Standard Analysis"
        case comprehensive = "Comprehensive Analysis"
        case comparison = "Comparison Analysis"
    }
}

struct VehicleInfo: Codable {
    let make: String
    let model: String
    let year: Int?
    let plateNumber: String
    let vin: String?
}

struct TyreInfo: Codable {
    let brand: String
    let model: String
    let size: String
    let dot: String
    let position: TyrePosition
    let season: String
    let loadIndex: String
    let speedRating: String

    enum TyrePosition: String, Codable {
        case frontLeft = "Front Left"
        case frontRight = "Front Right"
        case rearLeft = "Rear Left"
        case rearRight = "Rear Right"

        var shortCode: String {
            switch self {
            case .frontLeft: return "FL"
            case .frontRight: return "FR"
            case .rearLeft: return "RL"
            case .rearRight: return "RR"
            }
        }
    }
}

struct LocationInfo: Codable {
    let latitude: Double
    let longitude: Double
    let address: String?
}

struct WeatherInfo: Codable {
    let temperature: Double
    let humidity: Double
    let conditions: String
}

struct DepthMeasurementPoint: Codable, Identifiable {
    let id: UUID
    let x: Double  // Normalized 0-1
    let y: Double  // Normalized 0-1
    let depth: Double  // mm
    let confidence: Double  // 0-1
    let zone: TyreZone

    enum TyreZone: String, Codable {
        case center = "Center"
        case innerEdge = "Inner Edge"
        case outerEdge = "Outer Edge"
        case shoulder = "Shoulder"
    }

    init(id: UUID = UUID(), x: Double, y: Double, depth: Double, confidence: Double, zone: TyreZone) {
        self.id = id
        self.x = x
        self.y = y
        self.depth = depth
        self.confidence = confidence
        self.zone = zone
    }
}

struct WearCause: Codable, Identifiable {
    let id: UUID
    let type: CauseType
    let probability: Double  // 0-1
    let description: String

    enum CauseType: String, Codable {
        case overInflation = "Sovra-gonfiaggio"
        case underInflation = "Sotto-gonfiaggio"
        case misalignment = "Disallineamento"
        case improperRotation = "Mancata rotazione"
        case suspension = "Problemi sospensioni"
        case drivingStyle = "Stile di guida aggressivo"
        case braking = "Frenate brusche"
        case loading = "Carico eccessivo"
    }

    init(id: UUID = UUID(), type: CauseType, probability: Double, description: String) {
        self.id = id
        self.type = type
        self.probability = probability
        self.description = description
    }
}

// MARK: - Heat Map
struct DepthHeatMap: Codable {
    let gridSize: GridSize
    let dataPoints: [[Double]]  // 2D array of depth values
    let colorScheme: HeatMapColorScheme
    let interpolated: Bool

    struct GridSize: Codable {
        let rows: Int
        let columns: Int
    }

    enum HeatMapColorScheme: String, Codable {
        case rainbow = "Rainbow"
        case thermal = "Thermal"
        case grayscale = "Grayscale"
        case custom = "Custom"

        var colors: [Color] {
            switch self {
            case .rainbow:
                return [.red, .orange, .yellow, .green, .blue, .purple]
            case .thermal:
                return [.blue, .cyan, .green, .yellow, .orange, .red]
            case .grayscale:
                return [.black, .gray, .white]
            case .custom:
                return [Color(hex: "FF0000"), Color(hex: "FFFF00"), Color(hex: "00FF00")]
            }
        }
    }

    func colorForDepth(_ depth: Double, minDepth: Double = 0, maxDepth: Double = 8) -> Color {
        let normalized = (depth - minDepth) / (maxDepth - minDepth)
        let clampedValue = min(max(normalized, 0), 1)

        let colors = colorScheme.colors
        let index = Int(clampedValue * Double(colors.count - 1))
        return colors[min(index, colors.count - 1)]
    }
}

// MARK: - Remaining Life Estimate
struct RemainingLifeEstimate: Codable {
    let estimatedKilometers: Double
    let estimatedMonths: Int
    let confidence: Double  // 0-1
    let calculationMethod: CalculationMethod
    let factors: [LifeFactor]
    let projectedDepthCurve: [DepthProjection]

    enum CalculationMethod: String, Codable {
        case linear = "Linear Regression"
        case exponential = "Exponential Model"
        case machineLearning = "ML Prediction"
        case historical = "Historical Data"
    }

    var formattedDistance: String {
        if estimatedKilometers > 1000 {
            return String(format: "%.1f,000 km", estimatedKilometers / 1000)
        } else {
            return String(format: "%.0f km", estimatedKilometers)
        }
    }

    var status: LifeStatus {
        if estimatedKilometers > 20000 {
            return .excellent
        } else if estimatedKilometers > 10000 {
            return .good
        } else if estimatedKilometers > 5000 {
            return .fair
        } else if estimatedKilometers > 2000 {
            return .warning
        } else {
            return .critical
        }
    }

    enum LifeStatus {
        case excellent, good, fair, warning, critical

        var color: Color {
            switch self {
            case .excellent: return .green
            case .good: return .blue
            case .fair: return .yellow
            case .warning: return .orange
            case .critical: return .red
            }
        }
    }

    // Custom Codable to avoid requiring non-Codable element types for factors and projectedDepthCurve
    private enum CodingKeys: String, CodingKey {
        case estimatedKilometers
        case estimatedMonths
        case confidence
        case calculationMethod
        // Intentionally omit: factors, projectedDepthCurve
    }

    init(estimatedKilometers: Double,
         estimatedMonths: Int,
         confidence: Double,
         calculationMethod: CalculationMethod,
         factors: [LifeFactor],
         projectedDepthCurve: [DepthProjection]) {
        self.estimatedKilometers = estimatedKilometers
        self.estimatedMonths = estimatedMonths
        self.confidence = confidence
        self.calculationMethod = calculationMethod
        self.factors = factors
        self.projectedDepthCurve = projectedDepthCurve
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.estimatedKilometers = try container.decode(Double.self, forKey: .estimatedKilometers)
        self.estimatedMonths = try container.decode(Int.self, forKey: .estimatedMonths)
        self.confidence = try container.decode(Double.self, forKey: .confidence)
        self.calculationMethod = try container.decode(CalculationMethod.self, forKey: .calculationMethod)
        // Default to empty arrays when decoding since we intentionally do not decode these fields
        self.factors = []
        self.projectedDepthCurve = []
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(estimatedKilometers, forKey: .estimatedKilometers)
        try container.encode(estimatedMonths, forKey: .estimatedMonths)
        try container.encode(confidence, forKey: .confidence)
        try container.encode(calculationMethod, forKey: .calculationMethod)
        // Intentionally do not encode factors or projectedDepthCurve to avoid requiring their element types to conform to Codable
    }
}



// MARK: - Recommendations
struct Recommendation: Codable, Identifiable {
    let id: UUID
    let priority: Priority
    let category: Category
    let title: String
    let description: String
    let action: String
    let urgency: Urgency

    enum Priority: String, Codable {
        case critical = "Critical"
        case high = "High"
        case medium = "Medium"
        case low = "Low"

        var color: Color {
            switch self {
            case .critical: return .red
            case .high: return .orange
            case .medium: return .yellow
            case .low: return .blue
            }
        }

        var icon: String {
            switch self {
            case .critical: return "exclamationmark.triangle.fill"
            case .high: return "exclamationmark.circle.fill"
            case .medium: return "info.circle.fill"
            case .low: return "checkmark.circle.fill"
            }
        }
    }

    enum Category: String, Codable {
        case safety = "Safety"
        case maintenance = "Maintenance"
        case performance = "Performance"
        case cost = "Cost Optimization"
        case legal = "Legal Compliance"
    }

    enum Urgency: String, Codable {
        case immediate = "Immediate"
        case withinWeek = "Within a Week"
        case withinMonth = "Within a Month"
        case routine = "Routine Check"
    }

    init(id: UUID = UUID(), priority: Priority, category: Category, title: String, description: String, action: String, urgency: Urgency) {
        self.id = id
        self.priority = priority
        self.category = category
        self.title = title
        self.description = description
        self.action = action
        self.urgency = urgency
    }
}

// MARK: - Export Format
enum ReportExportFormat {
    case pdf
    case image(format: ImageFormat)
    case json
    case html

    enum ImageFormat {
        case png
        case jpeg(quality: Double)
    }

    var fileExtension: String {
        switch self {
        case .pdf: return "pdf"
        case .image(let format):
            switch format {
            case .png: return "png"
            case .jpeg: return "jpg"
            }
        case .json: return "json"
        case .html: return "html"
        }
    }
}

File: android/app/src/main/java/it/tyrevibes/app/core/model/TyreAnalysisReportModels.kt
package it.tyrevibes.app.core.model

import androidx.compose.ui.graphics.Color
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import java.util.UUID
import java.util.Date

// MARK: - Main Report Model
@Serializable
data class TyreAnalysisReport(
    val id: String = UUID.randomUUID().toString(),
    val metadata: ReportMetadata,
    @SerialName("depth_analysis") val depthAnalysis: DepthAnalysis,
    @SerialName("wear_analysis") val wearAnalysis: WearAnalysis,
    @SerialName("heat_map") val heatMap: DepthHeatMap,
    @SerialName("remaining_life") val remainingLife: RemainingLifeEstimate,
    val recommendations: List<Recommendation> = emptyList(),
    @SerialName("safety_score") val safetyScore: SafetyScore,
    val comparison: ComparisonData? = null
) {
    @Transient
    val createdAt: Long = metadata.timestamp
}

// MARK: - Report Metadata
@Serializable
data class ReportMetadata(
    @SerialName("report_id") val reportId: String,
    val timestamp: Long, // Unix timestamp in milliseconds
    val vehicle: VehicleInfo,
    val tyre: TyreInfo,
    val location: LocationInfo? = null,
    val weather: WeatherInfo? = null,
    @SerialName("analysis_type") val analysisType: AnalysisType
) {
    @Serializable
    enum class AnalysisType(val displayName: String) {
        @SerialName("quick") QUICK("Quick Scan"),
        @SerialName("standard") STANDARD("Standard Analysis"),
        @SerialName("comprehensive") COMPREHENSIVE("Comprehensive Analysis"),
        @SerialName("comparison") COMPARISON("Comparison Analysis")
    }
}

@Serializable
data class VehicleInfo(
    val make: String,
    val model: String,
    val year: Int? = null,
    @SerialName("plate_number") val plateNumber: String,
    val vin: String? = null
)

@Serializable
data class TyreInfo(
    val brand: String,
    val model: String,
    val size: String,
    val dot: String,
    val position: TyrePosition,
    val season: String,
    @SerialName("load_index") val loadIndex: String,
    @SerialName("speed_rating") val speedRating: String
) {
    @Serializable
    enum class TyrePosition(val displayName: String, val shortCode: String) {
        @SerialName("front_left") FRONT_LEFT("Front Left", "FL"),
        @SerialName("front_right") FRONT_RIGHT("Front Right", "FR"),
        @SerialName("rear_left") REAR_LEFT("Rear Left", "RL"),
        @SerialName("rear_right") REAR_RIGHT("Rear Right", "RR")
    }
}

@Serializable
data class LocationInfo(
    val latitude: Double,
    val longitude: Double,
    val address: String? = null
)

@Serializable
data class WeatherInfo(
    val temperature: Double,
    val humidity: Double,
    val conditions: String
)

@Serializable
data class DepthAnalysis(
    @SerialName("min_depth") val minDepth: Double,
    @SerialName("max_depth") val maxDepth: Double,
    @SerialName("avg_depth") val avgDepth: Double,
    @SerialName("measurement_points") val measurementPoints: List<DepthMeasurementPoint>
)

@Serializable
data class DepthMeasurementPoint(
    val id: String = UUID.randomUUID().toString(),
    val x: Double,  // Normalized 0-1
    val y: Double,  // Normalized 0-1
    val depth: Double,  // mm
    val confidence: Double,  // 0-1
    val zone: TyreZone
) {
    @Serializable
    enum class TyreZone(val displayName: String) {
        @SerialName("center") CENTER("Center"),
        @SerialName("inner_edge") INNER_EDGE("Inner Edge"),
        @SerialName("outer_edge") OUTER_EDGE("Outer Edge"),
        @SerialName("shoulder") SHOULDER("Shoulder")
    }
}

@Serializable
data class WearAnalysis(
    @SerialName("wear_pattern") val wearPattern: WearPattern,
    val severity: Double, // 0-1
    @SerialName("likely_causes") val likelyCauses: List<WearCause>
) {
    @Serializable
    enum class WearPattern(val displayName: String) {
        @SerialName("even") EVEN("Even Wear"),
        @SerialName("center") CENTER("Center Wear"),
        @SerialName("edge") EDGE("Edge Wear"),
        @SerialName("one_side") ONE_SIDE("One-Sided Wear"),
        @SerialName("cupping") CUPPING("Cupping"),
        @SerialName("feathering") FEATHERING("Feathering")
    }
}

@Serializable
data class WearCause(
    val id: String = UUID.randomUUID().toString(),
    val type: CauseType,
    val probability: Double,  // 0-1
    val description: String
) {
    @Serializable
    enum class CauseType(val displayName: String) {
        @SerialName("over_inflation") OVER_INFLATION("Sovra-gonfiaggio"),
        @SerialName("under_inflation") UNDER_INFLATION("Sotto-gonfiaggio"),
        @SerialName("misalignment") MISALIGNMENT("Disallineamento"),
        @SerialName("improper_rotation") IMPROPER_ROTATION("Mancata rotazione"),
        @SerialName("suspension") SUSPENSION("Problemi sospensioni"),
        @SerialName("driving_style") DRIVING_STYLE("Stile di guida aggressivo"),
        @SerialName("braking") BRAKING("Frenate brusche"),
        @SerialName("loading") LOADING("Carico eccessivo")
    }
}

// MARK: - Heat Map
@Serializable
data class DepthHeatMap(
    @SerialName("grid_size") val gridSize: GridSize,
    @SerialName("data_points") val dataPoints: List<List<Double>>,  // 2D array of depth values
    @SerialName("color_scheme") val colorScheme: HeatMapColorScheme,
    val interpolated: Boolean
) {
    @Serializable
    data class GridSize(
        val rows: Int,
        val columns: Int
    )

    @Serializable
    enum class HeatMapColorScheme {
        @SerialName("rainbow") RAINBOW,
        @SerialName("thermal") THERMAL,
        @SerialName("grayscale") GRAYSCALE,
        @SerialName("custom") CUSTOM;

        fun getColors(): List<Color> {
            return when (this) {
                RAINBOW -> listOf(
                    Color.Red, Color(0xFFFFA500), Color.Yellow,
                    Color.Green, Color.Blue, Color(0xFF800080)
                )
                THERMAL -> listOf(
                    Color.Blue, Color.Cyan, Color.Green,
                    Color.Yellow, Color(0xFFFFA500), Color.Red
                )
                GRAYSCALE -> listOf(Color.Black, Color.Gray, Color.White)
                CUSTOM -> listOf(Color(0xFFFF0000), Color(0xFFFFFF00), Color(0xFF00FF00))
            }
        }
    }

    fun colorForDepth(depth: Double, minDepth: Double = 0.0, maxDepth: Double = 8.0): Color {
        val normalized = (depth - minDepth) / (maxDepth - minDepth)
        val clampedValue = normalized.coerceIn(0.0, 1.0)

        val colors = colorScheme.getColors()
        val index = (clampedValue * (colors.size - 1)).toInt()
        return colors[index.coerceIn(0, colors.size - 1)]
    }
}

// MARK: - Remaining Life Estimate
@Serializable
data class RemainingLifeEstimate(
    @SerialName("estimated_kilometers") val estimatedKilometers: Double,
    @SerialName("estimated_months") val estimatedMonths: Int,
    val confidence: Double,  // 0-1
    @SerialName("calculation_method") val calculationMethod: CalculationMethod
) {
    @Transient
    val factors: List<LifeFactor> = emptyList()

    @Transient
    val projectedDepthCurve: List<DepthProjection> = emptyList()

    @Serializable
    enum class CalculationMethod(val displayName: String) {
        @SerialName("linear") LINEAR("Linear Regression"),
        @SerialName("exponential") EXPONENTIAL("Exponential Model"),
        @SerialName("ml") MACHINE_LEARNING("ML Prediction"),
        @SerialName("historical") HISTORICAL("Historical Data")
    }

    val formattedDistance: String
        get() = if (estimatedKilometers > 1000) {
            String.format("%.1f,000 km", estimatedKilometers / 1000)
        } else {
            String.format("%.0f km", estimatedKilometers)
        }

    val status: LifeStatus
        get() = when {
            estimatedKilometers > 20000 -> LifeStatus.EXCELLENT
            estimatedKilometers > 10000 -> LifeStatus.GOOD
            estimatedKilometers > 5000 -> LifeStatus.FAIR
            estimatedKilometers > 2000 -> LifeStatus.WARNING
            else -> LifeStatus.CRITICAL
        }

    enum class LifeStatus(val color: Color) {
        EXCELLENT(Color.Green),
        GOOD(Color.Blue),
        FAIR(Color.Yellow),
        WARNING(Color(0xFFFFA500)),
        CRITICAL(Color.Red)
    }
}

// Non-serializable supporting types
data class LifeFactor(
    val name: String,
    val impact: Double,
    val description: String
)

data class DepthProjection(
    val kilometers: Double,
    val expectedDepth: Double,
    val confidenceRange: DepthRange
)

data class DepthRange(
    val min: Double,
    val max: Double
)

@Serializable
data class SafetyScore(
    val overall: Double, // 0-100
    @SerialName("wet_performance") val wetPerformance: Double,
    @SerialName("dry_performance") val dryPerformance: Double,
    @SerialName("legal_compliance") val legalCompliance: Boolean,
    val grade: SafetyGrade
) {
    @Serializable
    enum class SafetyGrade(val displayName: String) {
        @SerialName("A") A("A - Excellent"),
        @SerialName("B") B("B - Good"),
        @SerialName("C") C("C - Fair"),
        @SerialName("D") D("D - Poor"),
        @SerialName("F") F("F - Dangerous")
    }
}

@Serializable
data class ComparisonData(
    @SerialName("previous_report_id") val previousReportId: String,
    @SerialName("days_elapsed") val daysElapsed: Int,
    @SerialName("depth_change") val depthChange: Double,
    @SerialName("wear_rate") val wearRate: Double
)

// MARK: - Recommendations
@Serializable
data class Recommendation(
    val id: String = UUID.randomUUID().toString(),
    val priority: Priority,
    val category: Category,
    val title: String,
    val description: String,
    val action: String,
    val urgency: Urgency
) {
    @Serializable
    enum class Priority(val displayName: String, val icon: String) {
        @SerialName("critical") CRITICAL("Critical", "warning"),
        @SerialName("high") HIGH("High", "error"),
        @SerialName("medium") MEDIUM("Medium", "info"),
        @SerialName("low") LOW("Low", "check_circle");

        fun getColor(): Color {
            return when (this) {
                CRITICAL -> Color.Red
                HIGH -> Color(0xFFFFA500)
                MEDIUM -> Color.Yellow
                LOW -> Color.Blue
            }
        }
    }

    @Serializable
    enum class Category(val displayName: String) {
        @SerialName("safety") SAFETY("Safety"),
        @SerialName("maintenance") MAINTENANCE("Maintenance"),
        @SerialName("performance") PERFORMANCE("Performance"),
        @SerialName("cost") COST("Cost Optimization"),
        @SerialName("legal") LEGAL("Legal Compliance")
    }

    @Serializable
    enum class Urgency(val displayName: String) {
        @SerialName("immediate") IMMEDIATE("Immediate"),
        @SerialName("within_week") WITHIN_WEEK("Within a Week"),
        @SerialName("within_month") WITHIN_MONTH("Within a Month"),
        @SerialName("routine") ROUTINE("Routine Check")
    }
}

// MARK: - Export Format
sealed class ReportExportFormat {
    object PDF : ReportExportFormat()
    data class Image(val format: ImageFormat) : ReportExportFormat()
    object JSON : ReportExportFormat()
    object HTML : ReportExportFormat()

    sealed class ImageFormat {
        object PNG : ImageFormat()
        data class JPEG(val quality: Double) : ImageFormat()
    }

    val fileExtension: String
        get() = when (this) {
            is PDF -> "pdf"
            is Image -> when (format) {
                is ImageFormat.PNG -> "png"
                is ImageFormat.JPEG -> "jpg"
            }
            is JSON -> "json"
            is HTML -> "html"
        }
}
